lib/Sema/CSGen.cpp

```diff
@@ -3543,16 +3543,36 @@ namespace {
     
     Type visitUnresolvedPatternExpr(UnresolvedPatternExpr *expr) {
       // Encountering an UnresolvedPatternExpr here means we have an invalid
-      // ExprPattern with a Pattern node like 'let x' nested in it. Record a
-      // fix, and assign ErrorTypes to any VarDecls bound.
-      auto *locator = CS.getConstraintLocator(expr);
+      // ExprPattern with a Pattern node like 'let x' nested in it. Assign
+      // ErrorTypes to any VarDecls bound, and record a fix after finding the
+      // correct anchor.
       auto *P = expr->getSubPattern();
-      CS.recordFix(IgnoreInvalidPatternInExpr::create(CS, P, locator));
-
       P->forEachVariable([&](VarDecl *VD) {
         CS.setType(VD, ErrorType::get(CS.getASTContext()));
       });
-
+      Expr *anchor = expr;
+      bool isInvalidCast = false;
+      // If we are in a CheckedCastExpr, it's possible that is the problem.
+      // E.g: `if case let x as? Int = y {}`. In which case, we want to anchor
+      // the fix there and set `isInvalidCast` = true.
+      auto *E = CS.getParentExpr(expr);
+      while (E && !isa<CheckedCastExpr>(E))
+        E = CS.getParentExpr(E);
+      auto *castExpr = cast_or_null<CheckedCastExpr>(E);
+      // Check if we are in an argument of `~=` that is _not_ inside a CallExpr,
+      // avoiding cases like: `case let f(x as? Int) = y {}`,
+      // where resolving the CheckedCastExpr would not be a fix.
+      auto *parent = castExpr ? CS.getParentExpr(castExpr) : nullptr;
+      while (parent && !isa<BinaryExpr>(parent))
+        parent = isa<CallExpr>(parent) ? nullptr : CS.getParentExpr(parent);
+      auto *BE = cast_or_null<BinaryExpr>(parent);
+      if (BE && isPatternMatchingOperator(BE->getFn())) {
+        anchor = castExpr;
+        isInvalidCast = true;
+      }
```

Comment:
I'd say all this logic here can go directly into `diagnoseInvalidCheckedCast`...

Reply:
Okay, the only issue I run into with it there is with tuples and emitting multiple of the same error. I was doing it here so I could put the anchor at the CheckedCastExpr and then since the locator is uniqued on that, we wouldn't get the multiple errors. However, maybe that was not a good way for me to go about it. Also, I may have misunderstood  your instruction, with the diagnostic -- is it better to create just a specialized function like here? Or a separate subclass of FailureDiagnostic?

Reply:
I will try moving the logic into the diagnostic function and then use `hasFixFor` potentially to see if we can skip diagnosing  a tuple.

Reply:
> is it better to create just a specialized function like diagnoseInvalidCheckedCast? Or a separate subclass of FailureDiagnostic?

Either way is fine, up to you.

Reply:
> I will try moving the logic into the diagnostic function and then use hasFixFor potentially to see if we can skip diagnosing a tuple.

Hm, it's the same diagnostic but at a different location in the expression because the fix gets anchored on each invalid pattern, right? I think that's fine.

Reply:
Oh, good to know, will remove that check then.